### PR TITLE
[NPUW] Enabled model with multiple outputs in LLMInferRequest

### PR DESCRIPTION
**Pull request by AsyaPronina**
### Details:
- *Added possibility for passed LLM model to have multiple outputs*

### Tickets:
 - *EISW-175831*

### CODE REVIEW

File: src/plugins/intel_npu/src/plugin/npuw/llm_infer_request.hpp [hunk 1]
```diff
@@ -10,6 +10,22 @@
...
+    std::size_t operator()(const ov::Output<const ov::Node>& port) const {
+        using std::hash;
+        using std::size_t;
+        using std::string;
+
+        // Compute individual hash values for first,
+        // second and third and combine them using XOR
+        // and bit shifting:
+
+        return ((hash<std::size_t>()(port.get_index()) ^ (hash<const ov::Node*>()(port.get_node()) << 1)) >> 1) ^
```
Comment (AlexanderKalistratov) [inline src/plugins/intel_npu/src/plugin/npuw/llm_infer_request.hpp:24]: Wouldn't it be enough just to have hash from `port.get_node()` ?

File: src/plugins/intel_npu/src/plugin/npuw/llm_infer_request.cpp [hunk 2]
```diff
@@ -274,16 +272,13 @@ void ov::npuw::LLMInferRequest::copy_kvcache() {
...
         const auto& output_name = kvcache_compiled->outputs()[i].get_any_name();
         auto prefill_out_tensor = m_prefill_request->get_tensor(m_prefill_out_ports.at(output_name));
 
         const auto& input_name = std::regex_replace(output_name, std::regex("present"), layer_names::past_key_values);
-        if (m_kvcache_in_ports.find(input_name) == m_kvcache_in_ports.end()) {
-            // FIXME: Totally wrong debug message. input_name is an invalid name of input layer.
-            LOG_DEBUG("Input name " << input_name << " doesn't contain kv cache. Skipping.");
-            continue;
-        }
+        NPUW_ASSERT(m_kvcache_in_ports.find(input_name) == m_kvcache_in_ports.end());
```
Comment (AlexanderKalistratov) [inline src/plugins/intel_npu/src/plugin/npuw/llm_infer_request.cpp:280]: This assert looks very suspicious

File: src/plugins/intel_npu/src/plugin/npuw/llm_compiled_model.hpp [hunk 3]
```diff
@@ -25,6 +25,11 @@ class LLMCompiledModel : public ov::npuw::ICompiledModel {
         uint32_t num_stored_tokens = 0u;
         uint32_t dim = 0u;
         bool v_tensors_transposed = false;
+        // TODO: As of now KV-cache blocks that are converted from states to
+        // model's I/O are appended to original model's I/O at the end,
+        // thus it is safe to loop over KVCache I/O blocks just using some
+        // start offsets.
+        std::size_t start_idx_in_outputs = 0u;
```
Comment (AlexanderKalistratov) [inline src/plugins/intel_npu/src/plugin/npuw/llm_compiled_model.hpp:32]: Where do we set this one to something different than 0?

File: src/plugins/intel_npu/src/plugin/npuw/llm_infer_request.cpp [hunk 4]
```diff
@@ -531,9 +530,9 @@ void ov::npuw::LLMInferRequest::infer_prefill(ov::SoPtr<ov::ITensor> input_ids,
     if (m_lm_head_request) {
         LOG_DEBUG("Calling inference for LM head model.");
         m_lm_head_request->infer();
-        m_logits = m_lm_head_request->get_tensor(m_lm_head_logits_port);
+        update_out_tensors_from(m_lm_head_request);
```
Comment (AsyaPronina) [inline src/plugins/intel_npu/src/plugin/npuw/llm_infer_request.cpp:533]: Wrong. Only `logits` will be in LM head, other outputs should be gathered from prefill.

File: src/plugins/intel_npu/src/plugin/npuw/llm_infer_request.hpp (hunk 1 above)
Comment (AsyaPronina) [reply src/plugins/intel_npu/src/plugin/npuw/llm_infer_request.hpp:24]: I am not sure, as `ov::Node` can have multiple outputs (with different indices)

File: src/plugins/intel_npu/src/plugin/npuw/llm_compiled_model.hpp (hunk 3 above)
Comment (AsyaPronina) [reply src/plugins/intel_npu/src/plugin/npuw/llm_compiled_model.hpp:32]: In constructor of `LLMCompiledModel`: https://github.com/openvinotoolkit/openvino/pull/31520/files#diff-7c0906d2d24827a3f37868e79470c1a41ea4bf66eae148a462c0442f8e9cc191R1087

File: src/plugins/intel_npu/src/plugin/npuw/llm_infer_request.hpp [hunk 5]
```diff
@@ -10,6 +10,22 @@
 #include "openvino/core/descriptor/output.hpp"
 #include "openvino/runtime/isync_infer_request.hpp"
 
+template <>
+struct std::hash<ov::Output<const ov::Node>> {
+    std::size_t operator()(const ov::Output<const ov::Node>& port) const {
+        using std::hash;
+        using std::size_t;
+        using std::string;
```
Comment (dmatveev) [inline src/plugins/intel_npu/src/plugin/npuw/llm_infer_request.hpp:18]: ???

**Review by dmatveev — COMMENTED** [review]
The solution here isn't radically better than the old one. You still make an assumption that you have kvcache tensors all starting at some position - instead of an explicit list you could detect/populate once and then use elsewhere.
